File: hbdb/hbdb/core/backend.py
import pickle
import struct
import threading
from typing import Dict, Any, Tuple, Optional, List, BinaryIO
from dataclasses import dataclass
from sortedcontainers import SortedDict
from .bloom import TableBloomFilter

# Snapshot format (shared with the C++ NativeBackend, see native/backend.cpp):
#   [Magic:4 "HBDB"][Version:4][NumKeys:8]
#   For each key:
#     [KeyLen:4][KeyBytes...]
#     [NumVers:4]
#     For each version (oldest first):
#       [TS:8][ValLen:4][ValPickledBytes...]
# The native code writes raw C structs, which are little-endian on every
# platform this builds for (x86-64/arm64), so "<" keeps files compatible
# across the Python and native backends.
_SNAP_MAGIC = b"HBDB"
_SNAP_VERSION = 1
_U32 = struct.Struct("<I")
_U64 = struct.Struct("<Q")


def _read_exact(f: BinaryIO, n: int) -> bytes:
    data = f.read(n)
    if len(data) != n:
        raise ValueError("Truncated snapshot file")
    return data


@dataclass
class VersionedValue:
    value: Any
    commit_ts: int

class VersionedKVStore:
    """
    Simulates the Storage Server role in FoundationDB.
    Stores multiple versions of keys (MVCC).
    """
    def __init__(self, force_python: bool = False):
        # Per-store filter: load_snapshot resets it, which must not wipe
        # entries belonging to other stores in the same process.
        self._bloom = TableBloomFilter()
        self._use_native = False

        if not force_python:
            try:
                from hbdb.native_ext import NativeBackend
                self._native = NativeBackend()
                self._use_native = True
                print("[HBDB] Using C++ NativeBackend 🚀")
            except ImportError:
                pass

        if not self._use_native:
            # Fallback to Python SortedDict
            self._store = SortedDict()
            self._lock = threading.RLock()

    def read(self, key: str, read_ts: int) -> Optional[Any]:
        if self._use_native:
            return self._native.read(key, read_ts)

        with self._lock:
            versions = self._store.get(key, [])
            for v in versions:
                if v.commit_ts <= read_ts:
                    return v.value
            return None

    def write(self, key: str, value: Any, commit_ts: int):
        # Update Bloom Filter (Python side)
        self._bloom.add(key)

        if self._use_native:
            self._native.write(key, value, commit_ts)
            return

        with self._lock:
            versions = self._store.setdefault(key, [])
            # Keep newest-first regardless of arrival order: WAL replay
            # and racing same-key commits can apply writes out of
            # timestamp order.
            i = 0
            while i < len(versions) and versions[i].commit_ts > commit_ts:
                i += 1
            versions.insert(i, VersionedValue(value, commit_ts))

    def scan(self, start_key: str, end_key: str, read_ts: int) -> List[Tuple[str, Any]]:
        if self._use_native:
            # Native returns list of (key, value). Filter None tombstones
            # (e.g. SQL DELETE) here for parity with the Python branch —
            # the C++ scan returns them verbatim.
            return [(k, v) for k, v in self._native.scan(start_key, end_key, read_ts)
                    if v is not None]

        results = []
        with self._lock:
            for key in self._store.irange(start_key, end_key, inclusive=(True, False)):
                val = self.read(key, read_ts)
                if val is not None:
                    results.append((key, val))
        return results

    def scan_keys(self, start_key: str, end_key: str) -> List[str]:
        """Return just keys in range (for index lookups)."""
        if self._use_native:
            # Native doesn't support scan_keys directly yet.
            # Only used for index lookups, which aren't in torture test hot path.
            # Hack: Scan with MAX_INT timestamp to get filtering? No, native scan filters by TS.
            # Using read_ts=2**64-1 (MAX_UINT64)
            full_scan = self._native.scan(start_key, end_key, 18446744073709551615)
            # Rebuild bloom filter on full scan or load? Ideally on load.
            return [k for k, v in full_scan]

        with self._lock:
            return list(self._store.irange(start_key, end_key, inclusive=(True, False)))

    def save_snapshot(self, path: str):
        if self._use_native:
            self._native.save_snapshot(path)
            return

        with self._lock:
            with open(path, "wb") as out:
                out.write(_SNAP_MAGIC)
                out.write(_U32.pack(_SNAP_VERSION))
                out.write(_U64.pack(len(self._store)))

                for key, versions in self._store.items():
                    kbytes = key.encode("utf-8")
                    out.write(_U32.pack(len(kbytes)))
                    out.write(kbytes)
                    out.write(_U32.pack(len(versions)))
                    # On disk versions are oldest-first (native append
                    # order); in memory we keep them newest-first.
                    for v in reversed(versions):
                        out.write(_U64.pack(v.commit_ts))
                        vbytes = pickle.dumps(v.value)
                        out.write(_U32.pack(len(vbytes)))
                        out.write(vbytes)

    def load_snapshot(self, path: str) -> int:
        if self._use_native:
            max_ts = self._native.load_snapshot(path)
            keys = (k for k, _ in self._native.scan("", "\xFF", 18446744073709551615))
        else:
            max_ts = self._py_load_snapshot(path)
            with self._lock:
                keys = list(self._store.keys())

        # Rebuild Bloom Filter from the restored keyspace
        self._bloom.reset()
        for k in keys:
            self._bloom.add(k)
        return max_ts

    def _py_load_snapshot(self, path: str) -> int:
        # Parse into a fresh store and swap at the end, so a corrupt or
        # truncated file raises without destroying the current state.
        new_store = SortedDict()
        max_ts = 0

        with open(path, "rb") as f:
            if _read_exact(f, 4) != _SNAP_MAGIC:
                raise ValueError("Invalid snapshot magic")
            (version,) = _U32.unpack(_read_exact(f, 4))
            if version != _SNAP_VERSION:
                raise ValueError(f"Unsupported snapshot version: {version}")
            (num_keys,) = _U64.unpack(_read_exact(f, 8))

            for _ in range(num_keys):
                (klen,) = _U32.unpack(_read_exact(f, 4))
                key = _read_exact(f, klen).decode("utf-8")
                (nver,) = _U32.unpack(_read_exact(f, 4))

                versions = []
                for _ in range(nver):
                    (ts,) = _U64.unpack(_read_exact(f, 8))
                    if ts > max_ts:
                        max_ts = ts
                    (vlen,) = _U32.unpack(_read_exact(f, 4))
                    value = pickle.loads(_read_exact(f, vlen))
                    versions.append(VersionedValue(value, ts))

                versions.reverse()  # disk is oldest-first; memory is newest-first
                new_store[key] = versions

        with self._lock:
            self._store = new_store

        return max_ts

File: hbdb/hbdb/native/backend.cpp
#include <algorithm>
#include <map>
#include <mutex>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <string>
#include <vector>

namespace py = pybind11;

struct VersionedValue {
  uint64_t ts;
  py::object value; // Store Python object directly
};

#include <fstream>
#include <iostream>

class NativeBackend {
public:
  NativeBackend() {}

  void write(const std::string &key, py::object value, uint64_t ts) {
    std::lock_guard<std::mutex> lock(mtx);
    auto &versions = store[key];
    // Keep ascending ts order regardless of arrival order: WAL replay
    // and racing same-key commits can apply writes out of timestamp
    // order. Common case (ascending arrivals) stays an append.
    auto it = versions.end();
    while (it != versions.begin() && std::prev(it)->ts > ts)
      --it;
    versions.insert(it, {ts, value});
  }

  py::object read(const std::string &key, uint64_t read_ts) {
    std::lock_guard<std::mutex> lock(mtx);
    auto it = store.find(key);
    if (it == store.end())
      return py::none();

    const auto &versions = it->second;
    // Find latest version <= read_ts
    // Versions are appended in order, so we can search backwards
    for (auto rit = versions.rbegin(); rit != versions.rend(); ++rit) {
      if (rit->ts <= read_ts) {
        return rit->value;
      }
    }
    return py::none();
  }

  std::vector<std::pair<std::string, py::object>>
  scan(const std::string &start, const std::string &end, uint64_t read_ts) {
    std::lock_guard<std::mutex> lock(mtx);
    std::vector<std::pair<std::string, py::object>> result;

    auto it = store.lower_bound(start);
    while (it != store.end()) {
      if (it->first >= end)
        break;

      const auto &versions = it->second;
      // Find visible version
      for (auto rit = versions.rbegin(); rit != versions.rend(); ++rit) {
        if (rit->ts <= read_ts) {
          result.push_back({it->first, rit->value});
          break;
        }
      }

      it++;
    }
    return result;
  }

  // Snapshot Format:
  // [Magic:4][Version:4][NumKeys:8]
  // Foreach Key:
  //   [KeyLen:4][KeyBytes...]
  //   [NumVers:4]
  //   Foreach Ver:
  //     [TS:8]
  //     [ValLen:4][ValPickledBytes...]

  void save_snapshot(const std::string &path) {
    std::lock_guard<std::mutex> lock(mtx);
    std::ofstream out(path, std::ios::binary);
    if (!out)
      throw std::runtime_error("Cannot open file for snapshot");

    const char magic[] = "HBDB";
    uint32_t version = 1;
    uint64_t num_keys = store.size();

    out.write(magic, 4);
    out.write(reinterpret_cast<char *>(&version), 4);
    out.write(reinterpret_cast<char *>(&num_keys), 8);

    py::object dumps = py::module::import("pickle").attr("dumps");

    for (const auto &kv : store) {
      const std::string &key = kv.first;
      const auto &versions = kv.second;

      uint32_t klen = key.size();
      out.write(reinterpret_cast<char *>(&klen), 4);
      out.write(key.data(), klen);

      uint32_t nver = versions.size();
      out.write(reinterpret_cast<char *>(&nver), 4);

      for (const auto &v : versions) {
        out.write(reinterpret_cast<const char *>(&v.ts), 8);

        // Serialize value using Python pickle
        py::bytes bytes = dumps(v.value);
        std::string s_bytes = bytes; // Copy to C++ string
        uint32_t vlen = s_bytes.size();

        out.write(reinterpret_cast<char *>(&vlen), 4);
        out.write(s_bytes.data(), vlen);
      }
    }
    out.close();
  }

  static void read_exact(std::ifstream &in, char *buf, std::streamsize n) {
    in.read(buf, n);
    if (in.gcount() != n)
      throw std::runtime_error("Truncated snapshot file");
  }

  uint64_t load_snapshot(const std::string &path) {
    std::lock_guard<std::mutex> lock(mtx);
    std::ifstream in(path, std::ios::binary);
    if (!in)
      throw std::runtime_error("Cannot open snapshot file");

    char magic[5] = {0};
    read_exact(in, magic, 4);
    if (std::string(magic) != "HBDB")
      throw std::runtime_error("Invalid snapshot magic");

    uint32_t version;
    read_exact(in, reinterpret_cast<char *>(&version), 4);
    if (version != 1)
      throw std::runtime_error("Unsupported snapshot version");

    uint64_t num_keys;
    read_exact(in, reinterpret_cast<char *>(&num_keys), 8);

    // Parse into a fresh map and swap at the end, so a corrupt or
    // truncated file throws without destroying the current state.
    std::map<std::string, std::vector<VersionedValue>> new_store;

    uint64_t max_ts = 0;

    py::object loads = py::module::import("pickle").attr("loads");

    for (uint64_t i = 0; i < num_keys; ++i) {
      uint32_t klen;
      read_exact(in, reinterpret_cast<char *>(&klen), 4);
      std::string key(klen, '\0');
      read_exact(in, &key[0], klen);

      uint32_t nver;
      read_exact(in, reinterpret_cast<char *>(&nver), 4);

      std::vector<VersionedValue> &versions = new_store[key];
      versions.reserve(nver);

      for (uint32_t j = 0; j < nver; ++j) {
        uint64_t ts;
        read_exact(in, reinterpret_cast<char *>(&ts), 8);
        if (ts > max_ts)
          max_ts = ts;

        uint32_t vlen;
        read_exact(in, reinterpret_cast<char *>(&vlen), 4);
        std::string vbytes(vlen, '\0');
        read_exact(in, &vbytes[0], vlen);

        // Deserialize
        py::object val = loads(py::bytes(vbytes));
        versions.push_back({ts, val});
      }
    }

    store.swap(new_store);
    return max_ts;
  }

private:
  std::mutex mtx;
  std::map<std::string, std::vector<VersionedValue>> store;
};

void init_backend(py::module &m) {
  py::class_<NativeBackend>(m, "NativeBackend")
      .def(py::init<>())
      .def("write", &NativeBackend::write)
      .def("read", &NativeBackend::read)
      .def("scan", &NativeBackend::scan)
      .def("save_snapshot", &NativeBackend::save_snapshot)
      .def("load_snapshot", &NativeBackend::load_snapshot);
}
